### PR TITLE
tcp: find bound device when laddr is ANY

### PR DESCRIPTION
**Pull request by zhhyu7**
icmp: find bound device when s_boundto is not zero

Signed-off-by: zhanghongyu <zhanghongyu@xiaomi.com>

## Summary
tcp: find bound device when laddr is ANY
icmp: find bound device when s_boundto is not zero

### CODE REVIEW

File: net/icmp/icmp_sendmsg.c [hunk 1]
```diff
@@ -316,7 +316,17 @@ ssize_t icmp_sendmsg(FAR struct socket *psock, FAR struct msghdr *msg,
 
   /* Get the device that will be used to route this ICMP ECHO request */
 
-  dev = netdev_findby_ripv4addr(INADDR_ANY, inaddr->sin_addr.s_addr);
+#ifdef CONFIG_NET_BINDTODEVICE
+  if (conn->sconn.s_boundto)
```
Comment (acassis) [inline net/icmp/icmp_sendmsg.c:320]:
```suggestion
  if (conn->sconn.s_boundto != 0)

File: net/tcp/tcp_finddev.c [hunk 2]
```diff
@@ -74,6 +75,11 @@ static int tcp_find_ipv4_device(FAR struct tcp_conn_s *conn,
 
   if (net_ipv4addr_cmp(addr, INADDR_ANY))
     {
+      if (local)
+        {
+          conn->dev = net_bound_device(&conn->sconn);
+        }
+
       return local ? OK : -EINVAL;
```
Comment (pkarashchenko) [inline net/tcp/tcp_finddev.c:83]:
```suggestion
      if (local)
        {
          conn->dev = net_bound_device(&conn->sconn);
          return OK;
        }

      return -EINVAL;
```

File: net/tcp/tcp_finddev.c [hunk 3]
```diff
@@ -124,6 +130,11 @@ static int tcp_find_ipv6_device(FAR struct tcp_conn_s *conn,
 
   if (net_ipv6addr_cmp(addr, g_ipv6_unspecaddr))
     {
+      if (local)
```
Comment (pkarashchenko) [inline net/tcp/tcp_finddev.c:133]: ditto

File: net/utils/net_bounddev.c [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        }
+    }
+
+  /* If no device was bound or the bound device is no longer valid,
+   * then let's try the default network device.
+   */
+
+  return dev == NULL ? netdev_default() : dev;
+}
+#endif
```
Comment (pkarashchenko) [inline net/utils/net_bounddev.c:84]:
```suggestion
#endif

```

File: net/utils/net_bounddev.c [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+       */
+
+      dev = netdev_findbyindex(sconn->s_boundto);
+      if (dev == NULL)
+        {
+          /* No device?  It must have been unregistered.  Un-bind the
+           * socket.
+           */
+
+          sconn->s_boundto = 0;
```
Comment (pkarashchenko) [inline net/utils/net_bounddev.c:74]:
```suggestion
          sconn->s_boundto = 0;
          return dev;
```

File: net/utils/net_bounddev.c [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        }
+    }
+
+  /* If no device was bound or the bound device is no longer valid,
+   * then let's try the default network device.
+   */
+
+  return dev == NULL ? netdev_default() : dev;
```
Comment (pkarashchenko) [inline net/utils/net_bounddev.c:82]:
```suggestion
  return netdev_default();
```

File: net/utils/net_bounddev.c (hunk 5 above)
Comment (zhhyu7) [reply net/utils/net_bounddev.c:74]: is line 75 should be return netdev_default? below return dev.

File: net/utils/net_bounddev.c (hunk 6 above)
Comment (zhhyu7) [reply net/utils/net_bounddev.c:82]: If sconn->s_boundto == 0, the semantics will be different. I just move this code from udp field to common, I wonder if this suggestion can be kept for now?

File: net/utils/net_bounddev.c (hunk 4 above)
Comment (pkarashchenko) [reply net/utils/net_bounddev.c:84]: Please add newline

**Review by pkarashchenko — COMMENTED** [review]
Please fix one leftover

**Review by pkarashchenko — APPROVED** [review]

**Pull request merged**